src/main/kotlin/com/github/vol0n/utbotcppclion/services/Client.kt

```diff
@@ -0,0 +1,270 @@
+package com.github.vol0n.utbotcppclion.services
+
+import com.github.vol0n.utbotcppclion.actions.AskServerToGenerateBuildDir
+import com.github.vol0n.utbotcppclion.actions.AskServerToGenerateJsonForProjectConfiguration
+import com.github.vol0n.utbotcppclion.actions.getProjectConfigRequestMessage
+import com.github.vol0n.utbotcppclion.actions.utils.notifyError
+import com.github.vol0n.utbotcppclion.actions.utils.notifyInfo
+import com.github.vol0n.utbotcppclion.actions.utils.notifyUnknownResponse
+import com.github.vol0n.utbotcppclion.client.GrpcStarter
+import com.github.vol0n.utbotcppclion.messaging.ConnectionStatus
+import com.github.vol0n.utbotcppclion.messaging.UTBotConnectionChangedNotifier
+import com.github.vol0n.utbotcppclion.ui.UTBotRequestProgressIndicator
+import com.github.vol0n.utbotcppclion.utils.createFileAndMakeDirs
+
+import testsgen.Testgen
+import testsgen.Util
+
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.project.Project
+
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.swing.Swing
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.withTimeout
+
+class Client(val project: Project) {
+    private var stub = GrpcStarter.startClient().stub
+    val grpcCoroutineScope: CoroutineScope
+    private var connectionStatus = ConnectionStatus.INIT
+    private val connectionChangedPublisher =
+        project.messageBus.syncPublisher(UTBotConnectionChangedNotifier.CONNECTION_CHANGED_TOPIC)
+    private var heartBeatJob: Job? = null
+    private val logger = Logger.getInstance(this::class.java)
+    private val metadata: io.grpc.Metadata = io.grpc.Metadata()
+
+    init {
+        val handler = CoroutineExceptionHandler { _, exception ->
+            exception.printStackTrace()
+        }
+        grpcCoroutineScope = CoroutineScope(Dispatchers.Swing + handler)
+
+        periodicHeartBeat()
+    }
+
+    fun getConnectionStatus() = connectionStatus
+    fun isServerAvailable() = connectionStatus == ConnectionStatus.CONNECTED
+
+    fun doHandShake() {
+        grpcCoroutineScope.launch {
+            try {
+                stub.handshake(Testgen.DummyRequest.newBuilder().build())
+            } catch (e: Exception) {
+                logger.warn("HandShake failed with the following error: ${e.message}")
+            }
+        }
+    }
+
+    private suspend fun setMetadata() {
+        val clientID = System.getenv("USER") ?: "someUserId"
```

Comment:
Well actually clientID is not user, but project description in UTBotCpp. And it needs to be unique. If server responds that it not unique, you should regenerate it with some random id. So I suggest to use project name and if it is not unique, try to add 5 random digits and letters